#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/data/supermarket_sales.csv')

In [4]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby("Branch")["gross income"].sum().sort_values(ascending=False)

,gross income
Branch,
C,5265.1765
A,5057.1605
B,5057.0320


Найприбутковіший філіал супермаркету С. У нього самий найбільший сумарний прибуток - 5265.1765.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [6]:
df.groupby(by=["Branch", "City"])["gross income"].sum().sort_values(ascending=False)

,,gross income
Branch,City,
C,Naypyitaw,5265.1765
A,Yangon,5057.1605
B,Mandalay,5057.0320


Філіал з найбільшим прибутком знаходиться у місті Naypyitaw.
У попередньому ДЗ ми розрахували, що у цьому місті найбільший середній чек і найбільше значення максимального чеку. Можливо, другий магазин поки не варто відкривати, а зосередитись поки на розвитку вже існуючого магазина, подумати, що можна вдосконалити.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [7]:
invoices_by_category = pd.pivot_table(df, index="Branch", columns="Product line", values="Invoice ID", aggfunc="size")

In [8]:
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [9]:
a = invoices_by_category["Electronic accessories"].max()

In [10]:
b = invoices_by_category['Electronic accessories'].idxmax()                 # Повертає назву філіалу (Branch) з найвищим значенням у колонці

In [11]:
print (f'У філії магазину {b} найбільше інвойсів із покупкою товарів категорії Electronic accessories: {a}')

У філії магазину A найбільше інвойсів із покупкою товарів категорії Electronic accessories: 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. Яка категорія товарів найбільш затребувана у філії міста Naypyitaw (за кількістю інвойсів)?

In [18]:
df[df['City'] == 'Naypyitaw'].groupby('Product line')['Invoice ID'].count().sort_values(ascending=False).head(1)

,Invoice ID
Product line,
Food and beverages,66


Найбільше користується попитом у місті Naypyitaw категорія товарів Food and beverages. Її можні масштабувати і розширювати.

2. Яка категорія товарів приносить найменший прибуток у філії міста Yangon?

In [21]:
df[df['City'] == 'Yangon'].groupby('Product line')['gross income'].sum().sort_values().head(1)

,gross income
Product line,
Health and beauty,599.893


Найменший дохід у місті Yangon приносить категорія товарів Healt and beauty. Потрібно переглянути асортимент наявний у цьому магазині, також ввести якісь промо-акції для стимулювання продажів даної категорії.

3. Який тип користувачів приносить найбільший дохід (постійні члени чи звичайні), який їх розподіл?

In [31]:
df.groupby("Customer type").agg({'gross income': 'sum', 'Invoice ID' : 'count'})

,gross income,Invoice ID
Customer type,,
Member,7820.164,501
Normal,7559.205,499


Постійні члени приносять трохи більший дохід і здійснюють трохи більше половини покупок. Можна зробити висновок, що програма лояльності працює. Потрібно залучати і простих користувачів в неї.